FIPS/fip-0032.md

```diff
@@ -21,290 +21,426 @@ replaces: N/A
-   gas model.
-2. The testnet plan laid out in FIP-0031.
+## Test Cases
+
+Refer to test plan laid out in [FIP-0031].
 
 ## Security Considerations
 
-TODO draft.
+Metering and pricing execution enables more precise gas accounting, which
+translates into higher network security.
+
+Note that this FIP alone is insufficient to secure the network for
+user-programmed actors; it is merely a stepping stone. A future FIP will
+introduce further the final gas model changes before enabling user
+programmability.
 
 ## Incentive Considerations
 
-With the introduction of execution gas to account for real execution cost, actor
-compute logic will no longer be exempted from gas charges, thus resulting in a
-net increase in gas expenditure in the network.
+The gas utilization of all messages will increase, except for value transfers
+which are unaffected.
+
+In order to quantify the magnitude of change, we have backtested this model
+against 2.5MM mainnet messages.
+
+- Annex A includes a detailed breakdown of net utilization increase per message
+  type (relative to original gas).
+- Annex B includes the final total gas per message type.
+
+On average, most message types increase their utilization between 1.5-3x. The
+gas performance of specific messages types is subject to various factors. The
+reader is advised to study Annex A to understand the variability involved.
+
+In our backtest, the four most important messages for storage provider
+operations were impacted in this manner:
+
+- SubmitWindowPoSt (miner actor): gas usage increases an average of 3x, with a
+  median of 2.73x, and a 90% percentile score of 4.42x. Some executions spiked
+  as much as 14x.
+- PreCommitSector (miner actor): gas usage increases an average of 3x, with a
+  median of 2.37x, and a 90% percentile score of 5x. Some executions spiked as
+  much as 9x.
+- ProveCommitSector (miner actor): gas usage increases an average of 1.21x, with
+  an identical median, and a 90% percentile score of 1.26x. Some executions
+  spiked as much as 2.7x.
+- PublishStorageDeals (market actor): gas usage increases an average of 1.71x,
+  with a median of 1.75x, and a 90% percentile score of 1.92x. Some executions
+  spiked as much as 2x.
+
+These two message types deserve special attention.
+
+- DeclareFaultsRecovered (miner actor): highest variability between executions
+  and highest multiplier; on average, gas cost increases 8x, with some instances
+  spiking as much as 30x in the backtest. All executions were within the block
+  gas limit.
+- ExtendSectorExpiration (miner actor): infrequent; was already an expensive
+  message. Some executions lead to gas usage beyoned the block limit, rendering
+  them inviable (up to 2.2e10 gas). However, this is a batched message, so
+  clients should size batches within gas limits.
+
+On average, we observe a global increase of 1.87x in gas utilization, if the
+current chain activity were to remain unchanged. However, the HyperDrive upgrade
```

Comment:
Making sure we are on the same page, this is saying total GasUsage for all mainnet transactions in the data dump will go up by 1.87x.

Reply:
Yes, the backtest indicates gas usage going up by 1.87x on average in the sample data.

Reply:
@vkalghatgi We do want to roll this technological change prior to enabling user programmability; this is part of the change management strategy. Rolling out a massive update where we basically change the operating system of the network AND we enable user programmability at the same time is tremedously risky.

On another note, user programmability will just deterioriate the situation _further_: messages for user actors will _add_ to miner messages, leading to even more demand pressure in the gas market, driving the base fee up.

Reply:
Curious, given the additional chain bandwidth in snarkpack aggregates that isn't currently being used, do we actually expect that "miner gas usage increases about 90% on average"? AFAIK, the batch balancer will kick in, and the increase will be much less in practice, right?

Reply:
> Curious, given the additional chain bandwidth in snarkpack aggregates that isn't currently being used, do we actually expect that "miner gas usage increases about 90% on average"? 

I think the 90% increase on average is given that the exact same set of messages are sent to the chain in the same exact epoch.

> AFAIK, the batch balancer will kick in, and the increase will be much less in practice, right?

Yes, there could be many other variables but in general we expect the total gas usage to be about the same as what it is now and BaseFee will hover around the network cross over.